Adding serializable enums (or enums with backing types and specified representations).

akeep opened the pull request:
Hey All,

Following up on the discussion in #611, this pull request attempts to capture the proposal in the P4-16 specification document.

jafingerhut commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2767,64 @@ X.v1  // reference to v1
+E e1 = E.e2;
+E e2;
+
+x  = (bit<8>) e1; // sets x to 1
```
Should that be "sets x to 0"?

jafingerhut replied:
I may be wrong in my comment, but note that it is fairly confusing to have variables with the exact same names as the enum member names, e.g. e1, e2 are used as both in this example code.

akeep replied:
No, e1 = E.e2, which is 1.  I should, however, chose different names for variables e1 and e2 (that was a really confusing choice on my part!)

jafingerhut commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2767,64 @@ X.v1  // reference to v1
+
+1. if an integer $x \notin N$ then $\forall e \in S$, `(E)x != e`
+1. $\forall x \in [0, 2^W]$, `(bit<W>)(E)x == x`
+1. $\forall e \in S$, `(E)(bit<W>)e == e`
```
So if I roughly translate this into English, if an enum type with a backing type `bit<W>`, where the numeric values defined by the enum definition do _not_ include all possible `2^W` values of the backing type, then it is guaranteed that variables of that enum type might not be equal to any of the named values in the enum type?

```
enum bit<2> MyPartialEnum_t {
    VALUE_A = 2w0,
    VALUE_B = 2w1,
    VALUE_C = 2w2
}

MyPartialEnum_t x;

if (x == MyPartialEnum_t.VALUE_A) {
    // some code here
} else if (x == MyPartialEnum_t.VALUE_B) {
    // some code here
} else if (x == MyPartialEnum_t.VALUE_C) {
    // some code here
} else {
    // A P4 compiler MUST ASSUME that this branch can be executed
    // some code here
}
```

akeep replied:
Should one (or both) of these examples be added as explanation to the math?

jamescoole-cisco replied:
It's not obvious that the first example is unique to enums with backing types. If uninitialized enum variables are intended to be "undefined" only within valid entries of the enum, it might be good to state that explicitly.

However, if that's the case, (compiler) implementations will already need to initialize non-backed enum variables in most cases, so why not require the same for enums with backing types?

jafingerhut replied:
@jamescoole-cisco I actually wrote a change to the P4_16 language spec, committed after v1.0.0 was published, that says something about exactly this case, but had forgotten I wrote that part.  See https://github.com/p4lang/p4-spec/issues/611#issuecomment-386702605 for details.  Unless that text is changed further in the spec, P4_16 compilers are already free to make uninitialized enum and error values unequal to any of their named members.

jafingerhut replied:
Given my previous comment, I am not against having example code like these in the spec, but if so, they should be clear that the examples apply for enum's with _or without_ underlying types.

I think having no code examples would also be fine, but adding a sentence that mentions that _all_ values with type enum can be not-equal to any of their named members, with a cross-reference to the section "## Operations on headers { #sec-ops-on-hdrs }" (Madoko syntax copied from the .mdk file)

jamescoole-cisco replied:
I agree it would be good to copy or move the discussion of how undefined both types of enums can be into section, since it applies in situations besides header fields. But IDK if the example is necessary to get the point across.

jamescoole-cisco commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2043,10 +2053,47 @@ enum Suits { Clubs, Diamonds, Hearths, Spades }
+for each entry in the `enum` and provides a backing type: `bit<16>`.
+Implementations are expected to raise an error if the numeric representation
+for an enumeration entry falls outside the representation range of the backing
+type.
```
It seems like the issue of multiple entries having the same representation should be discussed somewhere in here. For example, whether this is allowed:

```
enum bit<16> SomeEnum {
  Case1 = 0x0001,
  Case2 = 0x0001, // related, can I say instead "= Case1"?
  Case3 = 0x0002
}
```

jafingerhut replied:
Good point.  Given the contents of the rest of the proposal, it seems that this should not be allowed.

jamescoole-cisco commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2767,64 @@ X.v1  // reference to v1
+sets `e` to an unknown value, since there is no symbol corresponding to the
+numeric value `5`.  While the value of `e` is unspecified, it must not
+correspond to any symbol defined within enum `E`.  This is because casting from
+a numeric type to an `enum` type is required obey the following semantics:
```
Some of this has been discussed before and was in the proposal, but I'm not sure where it ended up.

It seems that providing something like `bool in_range<Enum, Back>(Back b)` (or `bool in_range<Enum>(Enum e)`), either as part of the language or an extern, would be pretty valuable.

If a user wants to check that an unsafe cast will (or did) succeed, they're still going to even if these aren't provided, but the code they write will need to be updated whenever the enum definition changes.

jamescoole-cisco commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2043,10 +2053,47 @@ enum Suits { Clubs, Diamonds, Hearths, Spades }
 target-specific).
 
+It is also possible to specify an `enum` with an underlying representation.
+This requires the programmer provide both the numeric type and an associated
```
Should "the numeric type" be "a fixed-width integer type"?

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2767,64 @@ X.v1  // reference to v1
+representations ($n$), the following properties must hold:
+
+1. if an integer $x \notin N$ then $\forall e \in S$, `(E)x != e`
+1. $\forall x \in [0, 2^W]$, `(bit<W>)(E)x == x`
```
I don' t know why you want this property to hold. Who cares about it?

antoninbas replied:
Do you mean to say the property holds by definition of an enum with underlying type and doesn't need to be spelled out explicitly?

mbudiu-vmw replied:
If that is the model, then we should say so explicitly in the spec.
This will preclude other implementations, and will also make it easier for people to understand what is going on. Do not leave freedom to the compiler writers if you think you will be sorry about it later. Like people designing C did with respect to integer overflow.

akeep replied:
This was requested in the issue #611 related to this and I basically pulled the text from there (with a lite-rewording & turning some text in math).  Where do we stand on this?

Personally, I don't feel strongly about it one way or the other.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2767,64 @@ X.v1  // reference to v1
+~ End P4Example
+
+sets `e` to an unknown value, since there is no symbol corresponding to the
+numeric value `5`.  While the value of `e` is unspecified, it must not
```
I think it is simpler to keep this to be unspecified, and make no such promises.

antoninbas replied:
Strongly disagree. I believe that would greatly limit how these enums can be used. Please see the parser example I gave below in a different comment.

akeep replied:
This property is important for programmers expectations. Without it the example that @antoninbas showed in the comments above is not something that a programmer could rely on working across P4 implementations.

akeep replied:
Ah, good catch, I missed the unspecified in this one.  That said, the `While the value ...` has now been removed.

mbudiu-vmw commented:
My proposal is to actually rephrase this to make it look more like a bit type with a bunch of constant values. Then the compiler does not have a choice: it will always implement such an enum with the underlying type. Also, the semantics of values that are not one of the constants is much simpler to describe. And you get all the properties that you are asking for for free.

So `enum bit<32> X { A = 1, B = 2 }` is in fact just a `bit<32>`, with a bunch of named constants X.A and X.B. Then we don't even have the requirement for A and B to be distinct, they can have any values you want. The compiler will treat them just as constants. Unlike a `bit<32>`, but much like a `newtype` (at least as it is today) this enum will have no arithmetic operations except tests for equality, assignments and casts from/to `bit<32>`. The most important gain is that now the semantics of a cast is very simple to describe: the same underlying bit representation is kept.

What do you think? So I think that the syntax is unchanged, you get all the properties you wanted to get, but it is much simpler for a programmer to understand what such an enum is doing. All the business about undefined values is gone.

antoninbas commented:
@mbudiu-vmw that works for me

akeep commented:
@antoninbas, @jamescoole-cisco, @jafingerhut I've added the examples with slight modification and moved the commentary around undefined values up into the operations section.

@mbudiu-vmw I've tried to add some sentences to the `enum` definition section to talk about it more like a `newtype` based on the `bit<W>` underlying type.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2832,6 +2840,77 @@ representations ($n$), the following properties must hold:
+
+Any `enum` that contains an unspecifed value, whether as the result of a cast
+to an `enum` with an underlying type, parse into the field of an `enum` with an
+underlying type, or simply the declaration of any `enum` without a specified
```
I think this explanation can now be also simplified by reducing it to the case of bitstrings.

akeep replied:
I'm not sure what you mean by this.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2076,6 +2076,14 @@ enum bit<16> EtherType {
 introduces a new enumeration type, which contains five constants---e.g.,
 `EtherType.IPV4`.  This `enum` declaration specifies the fixed-width integer representation
 for each entry in the `enum` and provides an underlying type: `bit<16>`.
+This type of `enum` declaration can be thought of as declaring a new `bit<16>`
```
Do we allow signed enum values as well?

akeep replied:
I've used unsigned in all of the examples, but I don't think there is any particular reason to restrict it to this.  Do you have an opinion?  (I'm happy to add a case for it.)

jafingerhut commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2775,163 @@ X.v1  // reference to v1
+E b;
+
+x = (bit<8>) a; // sets x to 1
+b = (E) x;      // sets e to E.e2
```
typo: "sets e" -> "sets b"

jafingerhut commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2775,163 @@ X.v1  // reference to v1
+sets `e` to an unknown value, since there is no symbol corresponding to the
+fixed-width integer value `5`.  While the value of `e` is unspecified, it must not
+correspond to any symbol defined within enum `E`.  This is because casting from
+a fixed-width integer type to an `enum` type is required to obey the following semantics:
```
I know you have the later condition that (bit<W>)(E)x == x for all possible `2^W` values of x of type `bit<W>`.  Assuming we keep that condition (seems reasonable to me), it isn't really true that the assignment to `e` in the example code above assigns `e` an "unknown" value?  I mean, it must be a value not equal to any named member of type `E`, I agree, but it must also be able to be cast back to a `bit<8>` type value of 5, and no other, so we know a lot more about it than that it is simply unknown.

I mention this simply because uninitialized values in P4_16 are, I believe, "more unknown" than the value of `e` in this sample code.

jamescoole-cisco commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2775,163 @@ X.v1  // reference to v1
+}
+~ End P4Example
+
+Note that each symbolic value in the enum must map to a unique fixed-width integer
```
Can case names be used while defining the enum?

This is particularly useful if aliases are allowed:

``` p4
enum bit<32> Authors {
  Clemens = 0,
  Twain = Clemens // is Clemens defined here? is Authors.Clemens?
}
```

mbudiu-vmw replied:
Not clear, a declaration is available only after it has been "completed". Not clear whether the Clemens declaration ends with the comma or with the closing brace.

jamescoole-cisco replied:
I'd argue that since it's useful, it should be allowed. But there should be some language to say one way or another.

akeep commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2775,163 @@ X.v1  // reference to v1
+use the specified underlying type as the serialization data type and
+representation.
+
+In addition to these operations a P4 implementation may choose to implement
```
I went ahead and removed these, I agree having them as part of an architecture description is probably right.

akeep commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2775,163 @@ X.v1  // reference to v1
+correspond to any symbol defined within enum `E`.  This is because casting from
+a fixed-width integer type to an `enum` type is required to obey the following semantics:
+
+Given an enum type $E$ with underlying type `bit<W>` and the set of mappings
```
I removed all of the language around this.

akeep commented:
Hey All,

I think I've addressed everything with the exception of whether an enum value should be matchable in a select as the underlying representation.  As mentioned in a comment above, I'm not a fan of that, but I've not added any text to address it yet.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2777,149 @@ X.v1  // reference to v1
+}
+~ End P4Example
+
+Additionally, within an `enum` declaration, earlier symbols may be referred to in later
```
Can we leave this out of the spec for now?
If anyone asks for this feature we should clarify.
But it will be easier to implement without it.

akeep replied:
Done.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2777,149 @@ X.v1  // reference to v1
+fixed-width integer representation for `E.e2`, `1`.  The variable `b` is then set to the
+symbolic value `E.e2`, which corresponds to the fixed-width integer value `1`.
+
+Note that while it is always safe to cast from an `enum` to its fixed-width integer type,
```
Maybe "safe" is not the right word here; with our current definition both casts are actually "safe", but the second one does not create a "named" value.

akeep replied:
Done.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2777,149 @@ X.v1  // reference to v1
+}
+~ End P4Example
+
+Any `enum` that contains an unnamed value, whether as the result of a cast
```
We should say here "variable with an enum type". The "enum" itself is the type declaration.

akeep replied:
Done.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2777,149 @@ X.v1  // reference to v1
+Any `enum` that contains an unnamed value, whether as the result of a cast
+to an `enum` with an underlying type, parse into the field of an `enum` with an
+underlying type, or simply the declaration of any `enum` without a specified
+initial value might not be equal to any of the values defined for that
```
Actually I think that "might" is too weak: it "will not" be equal for sure.

akeep replied:
I just changed "might" to "will", I think that is accurate.

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2717,13 +2777,149 @@ X.v1  // reference to v1
+to an `enum` with an underlying type, parse into the field of an `enum` with an
+underlying type, or simply the declaration of any `enum` without a specified
+initial value might not be equal to any of the values defined for that
+type.  Such an unspecified value should still lead to predictable
```
unspecified -> unnamed

akeep replied:
Done.

mbudiu-vmw commented on p4-16/spec/grammar.mdk:
```diff
@@ -334,6 +334,7 @@ structField
 
 enumDeclaration
     : optAnnotations ENUM name '{' identifierList '}'
+    | optAnnotations ENUM typeRef name '{' specifiedIdentifierList '}'
```
I have implemented this grammar change in the compiler and it will work out (using "bit" in fact) - there are no ambiguities.

akeep replied:
I've left this alone for now, if you prefer bit, let me know.

akeep commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -2026,12 +2026,22 @@ An enumeration type is defined using the following syntax:
 ~ Begin P4Grammar
 enumDeclaration
     : optAnnotations ENUM name '{' identifierList '}'
+    | optAnnotations ENUM typeRef name '{' specifiedIdentifierList '}'
```
You noted this here, but then noted later that the `typeRef` here worked, do you have a preference?

akeep commented:
Added `unsigned` to the description of the type you can use as the base type for a serializable `enum` type and updated the grammar to reflect this restriction.

akeep commented:
Hey All,

I'm not planning on making any additional changes, since I think I've addressed @mbudiu-vmw  requests and the requests that came up during the call on Monday.  If there are needed changes please let me know soon, I'll be traveling over the weekend and my opportunities to handle changes are going to be limited.
